File: src/lib/actions/rewards.ts
"use server";

import { and, eq, sql } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { db } from "@/db";
import { rewardRedemptions, rewards, users } from "@/db/schema";
import { requireActiveUser, requireAdmin } from "@/lib/auth/current-user";
import type { FormState } from "./form-state";

type RewardFieldsResult =
  | { error: string }
  | { title: string; description: string | null; pointCost: number };

function readRewardFields(formData: FormData): RewardFieldsResult {
  const title = String(formData.get("title") ?? "").trim();
  const description = String(formData.get("description") ?? "").trim();
  const pointCost = Number.parseInt(String(formData.get("pointCost") ?? "0"), 10);

  if (!title) return { error: "Title is required." };
  if (!Number.isFinite(pointCost) || pointCost <= 0) {
    return { error: "Point cost must be a positive number." };
  }

  return { title, description: description || null, pointCost };
}

export async function createReward(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const fields = readRewardFields(formData);
  if ("error" in fields) return { error: fields.error };

  await db.insert(rewards).values({
    id: crypto.randomUUID(),
    householdId: household.id,
    title: fields.title,
    description: fields.description,
    pointCost: fields.pointCost,
  });

  revalidatePath("/rewards");
  return null;
}

export async function updateReward(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const id = String(formData.get("id") ?? "");
  const existing = await db.query.rewards.findFirst({
    where: and(eq(rewards.id, id), eq(rewards.householdId, household.id)),
  });
  if (!existing) return { error: "Reward not found." };

  const fields = readRewardFields(formData);
  if ("error" in fields) return { error: fields.error };

  await db
    .update(rewards)
    .set({
      title: fields.title,
      description: fields.description,
      pointCost: fields.pointCost,
    })
    .where(eq(rewards.id, id));

  revalidatePath("/rewards");
  return null;
}

export async function toggleRewardActive(formData: FormData) {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const id = String(formData.get("id") ?? "");
  const existing = await db.query.rewards.findFirst({
    where: and(eq(rewards.id, id), eq(rewards.householdId, household.id)),
  });
  if (!existing) return;

  await db
    .update(rewards)
    .set({ isActive: !existing.isActive })
    .where(eq(rewards.id, id));

  revalidatePath("/rewards");
}

export async function deleteReward(formData: FormData) {
  const { household, user } = await requireActiveUser();
  requireAdmin(user);

  const id = String(formData.get("id") ?? "");
  await db
    .delete(rewards)
    .where(and(eq(rewards.id, id), eq(rewards.householdId, household.id)));

  revalidatePath("/rewards");
}

export async function redeemReward(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();

  const rewardId = String(formData.get("rewardId") ?? "");
  const reward = await db.query.rewards.findFirst({
    where: and(
      eq(rewards.id, rewardId),
      eq(rewards.householdId, household.id),
    ),
  });
  if (!reward || !reward.isActive) return { error: "Reward not available." };

  if (user.points < reward.pointCost) {
    return { error: "Not enough points yet." };
  }

  db.transaction((tx) => {
    tx.insert(rewardRedemptions)
      .values({
        id: crypto.randomUUID(),
        rewardId: reward.id,
        userId: user.id,
        pointsSpent: reward.pointCost,
      })
      .run();

    tx.update(users)
      .set({ points: sql`${users.points} - ${reward.pointCost}` })
      .where(eq(users.id, user.id))
      .run();
  });

  revalidatePath("/rewards");
  revalidatePath("/");
  return null;
}

File: backend/internal/api/handlers/rewards.go
package handlers

import (
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"

	"hestia/backend/internal/api/middleware"
	"hestia/backend/internal/services"
)

func (d *Deps) ListRewards(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	rewards, err := d.Reward.List(household.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list rewards"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"rewards": rewards})
}

type rewardRequest struct {
	Title       string  `json:"title"`
	Description *string `json:"description"`
	PointCost   int     `json:"pointCost"`
}

func (r *rewardRequest) toInput() (services.RewardInput, string) {
	if r.Title == "" {
		return services.RewardInput{}, "title is required"
	}
	if r.PointCost <= 0 {
		return services.RewardInput{}, "point cost must be a positive number"
	}
	return services.RewardInput{Title: r.Title, Description: r.Description, PointCost: r.PointCost}, ""
}

func (d *Deps) CreateReward(c *gin.Context) {
	household := middleware.CurrentHousehold(c)

	var req rewardRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
		return
	}
	input, errMsg := req.toInput()
	if errMsg != "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	reward, err := d.Reward.Create(household.ID, input)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create reward"})
		return
	}
	c.JSON(http.StatusCreated, reward)
}

func (d *Deps) UpdateReward(c *gin.Context) {
	household := middleware.CurrentHousehold(c)

	var req rewardRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
		return
	}
	input, errMsg := req.toInput()
	if errMsg != "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": errMsg})
		return
	}

	reward, err := d.Reward.Update(household.ID, c.Param("id"), input)
	if err != nil {
		if errors.Is(err, services.ErrNotFound) {
			c.JSON(http.StatusNotFound, gin.H{"error": "reward not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to update reward"})
		return
	}
	c.JSON(http.StatusOK, reward)
}

func (d *Deps) ToggleRewardActive(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	reward, err := d.Reward.ToggleActive(household.ID, c.Param("id"))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "reward not found"})
		return
	}
	c.JSON(http.StatusOK, reward)
}

func (d *Deps) DeleteReward(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	if err := d.Reward.Delete(household.ID, c.Param("id")); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete reward"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

func (d *Deps) RedeemReward(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	user := middleware.CurrentUser(c)
	rewardID := c.Param("id")

	redemption, err := d.Reward.Redeem(household.ID, rewardID, user.ID, user.Points)
	if err != nil {
		switch {
		case errors.Is(err, services.ErrNotFound), errors.Is(err, services.ErrRewardUnavailable):
			c.JSON(http.StatusBadRequest, gin.H{"error": "reward not available"})
		case errors.Is(err, services.ErrInsufficientPoints):
			c.JSON(http.StatusBadRequest, gin.H{"error": "not enough points yet"})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to redeem reward"})
		}
		return
	}

	d.broadcastEvent("reward.redeemed", map[string]any{"rewardId": rewardID, "userId": user.ID})
	c.JSON(http.StatusOK, redemption)
}
